Fix #6617: fixing startup NPE and synchronization on mListenerMasks

svatoun opened the pull request:
See issue https://github.com/JMRI/JMRI/issues/6617

The simplest thing is to make the constructor-initialized fields `final` - according to JLS, this will make the values immediately visible **after** the object is initialized (last constructor finishes). 

The current code also does not access the `mListenerMasks` and other fields outside the object. Making it private would ensure that noone will bypass synchronization requirements.

Next, while `addXNetListener` is synchronized, the access to `mListenerMasks` in forwardReply is not: it may read even inconsistent state of the HashMap, if not all of the data flushed; there's no shared monitor/volatile guard that the writing + reading threads would be ordered by in the `happens-before` relation. Synchronizing on `this`.

I would advise against using a ConcurrentHashMap: unless the r/w contention is really high, the CHM has lower performance than a simple synchronized. Unless this need to be really exposed to clients or subclasses, Collections.synchronizedMap() would add an unnecessary level of indirection; again plain synchronized may be better.

Nitpick: the NPE resulting from unboxing `null` from the Map may be avoided at all - since NPE handler corresponds to the `XNetInterface.ALL` branch, `getOrDefault` can be used to unify the processing.

Removing the NPE handler since:
- real NPEs will surface soon
- expected null processing can be written without exception and `if`
- will uncover potential concurrency issues

The are constraints not only the individual collections, but across them as well: i.e. code calling `pollMessage` assumes that it will get the corresponding listener from `pollReplyHandler`. Imagine a thread But IF two threads some to `sendHighPriorityXNetMessage` at the same time, the concurrent execution `highPriorityListeners.put` and `highPriorityQueue.put` may mix message/reply pairs in the queue order. Even worse, blocking asusmption on the listener queue would fail, as there could be the "other" listener already. 

Now imagine the processor for the message and `sendHighPriorityXNetMessage` would run really concurrently; the reader may get the message, and then execute `highPriorityListeners.isEmpty` before it gets populated by the writer. Unlikely, but remember Murphy's laws.


Nitpick: reader reads the queues in the order: message, listener. In highly concurrent environment (not the case I think), this will be the total sequence of ops
1. thread A puts message
2. thread B reads message
3. thread B blocks (will not, it will just fail to deliver reply) at listener queue, which is empty
4. thread A puts listener
5. thread B ublock, gets the listener and processes message-reply

The block at (3) can be simply avoided by writing to the message queue **last**, so the reader's first operation will either recognize there are no messages and do nothing; or there will be a message AND the listener will be already prepared for the reader.

Only here I have discovered, that although the queues are BlockingQueues the blocking feature is completely killed by the surrounding implementation. Blocking behaviour of `highPriorityQueue` is killed - the queue is not size-bounded, so it won't ever block on `put`; `offer` can be used to avoid unnecessary `catch` block. Since the reader first checks for emptiness, the queue won't ever (intentionally) block on `take`. It can block (and cause a bug) when two concurrent threads processes the queue, but that's not the case, I believe. 
Note: instead `offer`, `add` can be conservatively used as it will throw exception on reached capacity - for the case someone will change the code in the future and applies size-limit.

It's really a question why the hevyweight, highly concurrent (and therefore slower) queue impls should be used instead of plain collections. Currently it just synchronizes sequence of `isEmpty` + `take` against `put`; more simpler `Collections.synchronizedXXX` would be sufficient or even more primitive sync as the collections themselves need not (?) to be exposed to clients as API.

The `null` value of `pollReplyHandler` is strictly useful just in tests, the production code would malfunction if the function ever returned `null` (not call the listener it was posted with the message).

Anyway, leaving BlockingQueues as they are, they're misused, but doe not actually break anything except performance.

Note: the two queues actually carry a pair which is to be processed as a logical item. Instead of 2 coordinated queues, one queue w/ typed lightweight pairs/tuples can be used to better encode the intent. I didn't want to change APIs just to fix a bug.

---
The other stuff are trivial things discovered as I swept the surrounding code. I did not change them - if it will be agreed on that the comments are valid / worth the change I'll extend the fix otherwise remove comemnts.

Note that similar patterns are found in `DCCppTrafficController` and possibly elsewhere. Once agreed on approach, will search & apply on other places.

svatoun commented on java/src/jmri/jmrix/lenz/XNetTrafficController.java:
```diff
@@ -20,7 +20,9 @@
-    protected HashMap<XNetListener, Integer> mListenerMasks;
+    // @GuardedBy(this)
+    // PENDING: the field should be probably made private w/ accessor to force proper synchronization for reading.
+    protected final HashMap<XNetListener, Integer> mListenerMasks;
```
thanks. I need to find out why Find Usages do not report those usages to me.

svatoun replied:
re. GuardedBy: sorry - I overlooked the real annotation can be used - great ! I am used to mark variables from projects which could not use such checker ;) Definitely better with real annotation + check here.

pabender replied:
It is also possible that I streamlined the behavior over the years and these were at one time accessed in a child class but no longer are.

The DCCpp code started life as a clone of the XPressNet code.  Someone took us too literally when we said “use another system as a model”.  It has many sections of the XPressNet code that are commented out or just not used still in place ( I have been cleaning some of that up as time permits ).

svatoun replied:
Annotation added; pls. confirm the access modifiers. It's always better not to expose anything unless necessary ;)

pabender replied:
the original code had this without the quotes.  You modified that to be a string containing this which is wrong.  Please return that to the original form.

svatoun replied:
```
JMRI/java/src/jmri/jmrix/lenz/XNetTrafficController.java:24: error: incompatible types: XNetTrafficController cannot be converted to String
    @GuardedBy(this)
```

pabender replied:
Yes, that is the right one to use.

I guess we should update our annotation page

svatoun replied:
OK, so when leaving the annotation String-valued; do you see some obvious error why, if I remove `synchronized` around the `mListenerMasks` access in code, spotbugs did not produce a warning ?
Thanks.

svatoun commented on java/src/jmri/jmrix/lenz/XNetTrafficController.java:
```diff
@@ -90,67 +92,65 @@ public void forwardReply(AbstractMRListener client, AbstractMRReply m) {
-        }
+    // PENDING: these fields should be probably made private w/ accessor to force proper synchronization for reading.
+    final LinkedBlockingQueue<XNetMessage> highPriorityQueue;
+    final LinkedBlockingQueue<XNetListener> highPriorityListeners;
```
Question: if I understand @GuardedBy correctly, then adding the annotation would require to access those queues strictly under a monitor. But, as long as the processor is single-threaded (and it is - please correct if that's wrong assumption), `pollMessage` can just omit the locks; internal state of BlockingQueue is thread-safe and since only a single thread consumes stuff, no lock is necessary. The same applies to `pollReplyHandler`.
Just writing to the two has to be atomic, which @GuardedBy syntax cannot describe (IMHO)

pabender replied:
You are making the assumption that the processor is single threaded.  I don’t think we can or should do that.

This is really a producer consumer queue with multiple producers.  There may only be one consumer.

svatoun replied:
By 'processor' I meant the consumer, in your terminology; sorry for the confusion. How many threads call `pollMessage` to extract items from the queue ?

pabender replied:
In theory this only one thread removes items from the queue ( the single consumer ).  Multiple threads can place items in the queue, and that should all be synchronized.

svatoun replied:
One consumer. Understood.

I suppose I could easily do some error in the implications - so if you are willing to discuss, let's verify piece by piece. Even if the state would be OK as it is, a separate decision should define whether for the sake of readability a redundant, but immediately visible, synchronization should be added. If you feel your time is being wasted at any time, please tell me; the I am benefiting from the discussion (education, excercise on abstract example) more than the project.

First the consumer side
. 
**Accessing fields**: since the fields are now `final`, then according to JLS 17.5 ("Final field semantics"), after `XNetTrafficController` constructor finishes, any thread that can observe a reference value A of `XNetTrafficController` will observe correct (initialized) values of those final fields in A. So no sycnhronization is needed to observe initialized values.

**Accesing shared state**: since each op on `LinkedBlockingQueue` is thread-safe, no extra synchronization is needed to assure internal consistency.

**Consistency of operation sequence**: According to `LinkedBlockingQueue` semantics, each single operation is atomic and leaves a consistent state visible to everyone. Once `isEmpty` becomes true (after atomic put/offer completes), it stays consistent with contents of the queue because of internal synchronization within LinkedBlockingQueue. If an item would be added (concurrently) to the FIFO queue in between `isEmpty` and `take` executed by consumer (in this order), it will not affect the enumeration order of `take`s thanks to internal synchronization of the `LinkedBlockingQueue`.
So plain check + reading ONE of the queues should not require locking - as long as the all the consumer calls run in one and single thread.

Is the above OK ?

svatoun commented:
I have changed `offer` -> `add`: offer produces spotbugs warnings on ignoring `boolean` return value.

I wonder if I run `spotbugs` right - I've added `@GuardedBy` to `mListenerMasks` and removed the synchronized block in `forwardReply` ... but no warnings showed up in the spotbugs report.

Run ```ant -Dspotbugs.home=/space/apps/spotbugs/spotbugs-3.1.11/ clean spotbugs``` in JMRI root.

bobjacobsen commented:
@pabender I haven't been following all the details of this. Have the changes you requested been made?

pabender commented:
@bobjacobsen The annotation was fixed and the access mode restored, but there are still a couple of comments in the code that should be removed.  That isn’t something that should keep this from getting merged however.

svatoun commented:
@pabender  on the other hand, it's nothing which has to be changed/fixed ASAP, given that it has been there for some time with (maybe ?) just occasional failure. If you do not feel comfortable with the changes, maybe they should be tuned before merge. You call.

pabender commented:
@svatoun I am OK with the code changes as they are now... with the caveat that I still have not tried this with real hardware.
